In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

In [2]:
df=pd.read_csv("survey_results.csv")
df.head()


Timestamp   Age Primary streaming service  Hours per day  \
0  8/27/2022 19:29:02  18.0                   Spotify            3.0   
1  8/27/2022 19:57:31  63.0                   Pandora            1.5   
2  8/27/2022 21:28:18  18.0                   Spotify            4.0   
3  8/27/2022 21:40:40  61.0             YouTube Music            2.5   
4  8/27/2022 21:54:47  18.0                   Spotify            4.0   

  While working Instrumentalist Composer         Fav genre Exploratory  \
0           Yes             Yes      Yes             Latin         Yes   
1           Yes              No       No              Rock         Yes   
2            No              No       No  Video game music          No   
3           Yes              No      Yes              Jazz         Yes   
4           Yes              No       No               R&B         Yes   

  Foreign languages  ...  Frequency [R&B]  Frequency [Rap] Frequency [Rock]  \
0               Yes  ...        Sometimes  Very frequently            Never   
1                No  ...        Sometimes           Rarely  Very frequently   
2               Yes  ...            Never           Rarely           Rarely   
3               Yes  ...        Sometimes            Never            Never   
4                No  ...  Very frequently  Very frequently            Never   

  Frequency [Video game music] Anxiety Depression Insomnia  OCD Music effects  \
0                    Sometimes     3.0        0.0      1.0  0.0           NaN   
1                       Rarely     7.0        2.0      2.0  1.0           NaN   
2              Very frequently     7.0        7.0     10.0  2.0     No effect   
3                        Never     9.0        7.0      3.0  3.0       Improve   
4                       Rarely     7.0        2.0      5.0  9.0       Improve   

     Permissions  
0  I understand.  
1  I understand.  
2  I understand.  
3  I understand.  
4  I understand.  

[5 rows x 33 columns]

In [3]:
df["Anxiety"].value_counts()

Anxiety
7.0     122
8.0     115
6.0      83
3.0      69
10.0     67
5.0      59
9.0      56
4.0      56
2.0      44
0.0      35
1.0      29
7.5       1
Name: count, dtype: int64

In [4]:
# 7.5 convert to 7 as it is only one value in the column
df[df["Anxiety"]==7.5]=8.0


In [5]:
#similarly

df["Depression"].value_counts()
df[df["Anxiety"]==3.5]=4

In [6]:
df["Insomnia"].value_counts()
df[df["Anxiety"]==3.5]=4

In [7]:
df["OCD"].value_counts()
df[df["Anxiety"]==5.5]=6

In [8]:
# suming all mental indexes
df["mental_scale"]=(df["Anxiety"]+df["Depression"]+df["Insomnia"]+df["OCD"])/4



Respondents rank Anxiety, Depression, Insomnia, and OCD on a scale of 0 to 10, where:

0 - I do not experience t  s.
10 - I experience this regularly, constantly/or to an extreme.

In [9]:
# cleaning and eda

In [10]:
# droping timestamp and permission
df=df.drop(columns=["Timestamp","Permissions","Anxiety","Depression","Insomnia","OCD"],axis=1)
df

Age Primary streaming service  Hours per day While working  \
0    18.0                   Spotify            3.0           Yes   
1    63.0                   Pandora            1.5           Yes   
2    18.0                   Spotify            4.0            No   
3    61.0             YouTube Music            2.5           Yes   
4    18.0                   Spotify            4.0           Yes   
..    ...                       ...            ...           ...   
731  17.0                   Spotify            2.0           Yes   
732  18.0                   Spotify            1.0           Yes   
733  19.0   Other streaming service            6.0           Yes   
734  19.0                   Spotify            5.0           Yes   
735  29.0             YouTube Music            2.0           Yes   

    Instrumentalist Composer         Fav genre Exploratory Foreign languages  \
0               Yes      Yes             Latin         Yes               Yes   
1                No       No              Rock         Yes                No   
2                No       No  Video game music          No               Yes   
3                No      Yes              Jazz         Yes               Yes   
4                No       No               R&B         Yes                No   
..              ...      ...               ...         ...               ...   
731             Yes       No              Rock         Yes               Yes   
732             Yes       No               Pop         Yes               Yes   
733              No      Yes               Rap         Yes                No   
734             Yes       No         Classical          No                No   
735              No       No           Hip hop         Yes               Yes   

       BPM  ... Frequency [Latin] Frequency [Lofi] Frequency [Metal]  \
0    156.0  ...   Very frequently           Rarely             Never   
1    119.0  ...         Sometimes           Rarely             Never   
2    132.0  ...             Never        Sometimes         Sometimes   
3     84.0  ...   Very frequently        Sometimes             Never   
4    107.0  ...         Sometimes        Sometimes             Never   
..     ...  ...               ...              ...               ...   
731  120.0  ...         Sometimes           Rarely            Rarely   
732  160.0  ...             Never           Rarely             Never   
733  120.0  ...            Rarely        Sometimes            Rarely   
734  170.0  ...             Never            Never             Never   
735   98.0  ...             Never           Rarely             Never   

     Frequency [Pop]  Frequency [R&B]  Frequency [Rap] Frequency [Rock]  \
0    Very frequently        Sometimes  Very frequently            Never   
1          Sometimes        Sometimes           Rarely  Very frequently   
2             Rarely            Never           Rarely           Rarely   
3          Sometimes        Sometimes            Never            Never   
4          Sometimes  Very frequently  Very frequently            Never   
..               ...              ...              ...              ...   
731  Very frequently            Never           Rarely  Very frequently   
732  Very frequently            Never            Never        Sometimes   
733        Sometimes        Sometimes        Sometimes           Rarely   
734            Never            Never            Never            Never   
735        Sometimes  Very frequently  Very frequently  Very frequently   

    Frequency [Video game music] Music effects mental_scale  
0                      Sometimes           NaN         1.00  
1                         Rarely           NaN         3.00  
2                Very frequently     No effect         6.50  
3                          Never       Improve         5.50  
4                         Rarely       Improve         5.75  
..                           ...           ...          ...  
731                        N

In [11]:
df.head()
df.shape

(736, 28)

In [12]:
df["Primary streaming service"].value_counts()
df[df["Primary streaming service"]==8.0] 
# tf is this all 8 .droping this

Age Primary streaming service  Hours per day While working  \
430  8.0                       8.0            8.0           8.0   

    Instrumentalist Composer Fav genre Exploratory Foreign languages  BPM  \
430             8.0      8.0       8.0         8.0               8.0  8.0   

     ... Frequency [Latin] Frequency [Lofi] Frequency [Metal] Frequency [Pop]  \
430  ...               8.0              8.0               8.0             8.0   

    Frequency [R&B] Frequency [Rap] Frequency [Rock]  \
430             8.0             8.0              8.0   

    Frequency [Video game music] Music effects mental_scale  
430                          8.0           8.0          8.0  

[1 rows x 28 columns]

In [13]:
df=df.drop(index=430,axis=0) 
df[df["Primary streaming service"]==8.0] 

Empty DataFrame
Columns: [Age, Primary streaming service, Hours per day, While working, Instrumentalist, Composer, Fav genre, Exploratory, Foreign languages, BPM, Frequency [Classical], Frequency [Country], Frequency [EDM], Frequency [Folk], Frequency [Gospel], Frequency [Hip hop], Frequency [Jazz], Frequency [K pop], Frequency [Latin], Frequency [Lofi], Frequency [Metal], Frequency [Pop], Frequency [R&B], Frequency [Rap], Frequency [Rock], Frequency [Video game music], Music effects, mental_scale]
Index: []

[0 rows x 28 columns]

In [14]:

df.isna().sum()

Age                               1
Primary streaming service         1
Hours per day                     0
While working                     3
Instrumentalist                   4
Composer                          1
Fav genre                         0
Exploratory                       0
Foreign languages                 4
BPM                             107
Frequency [Classical]             0
Frequency [Country]               0
Frequency [EDM]                   0
Frequency [Folk]                  0
Frequency [Gospel]                0
Frequency [Hip hop]               0
Frequency [Jazz]                  0
Frequency [K pop]                 0
Frequency [Latin]                 0
Frequency [Lofi]                  0
Frequency [Metal]                 0
Frequency [Pop]                   0
Frequency [R&B]                   0
Frequency [Rap]                   0
Frequency [Rock]                  0
Frequency [Video game music]      0
Music effects                     8
mental_scale                

this na values is handle later by simple inputer, here  Anxiety,Depression
Insomia ,ocd 
 has no null value 

In [15]:
df["BPM"].sort_values(ascending=False)

568    999999999.0
644          624.0
248          220.0
610          220.0
662          218.0
          ...     
688            NaN
700            NaN
706            NaN
712            NaN
717            NaN
Name: BPM, Length: 735, dtype: float64

the BPM is so random ig , what is 999999999. i think surveyee randomly put those value.

In [16]:
df.iloc[568:569]
# i think droping this value is for the best
df=df.drop(568)

In [17]:
# checking if there are any more those values
df[df['BPM'] > 1000]

Empty DataFrame
Columns: [Age, Primary streaming service, Hours per day, While working, Instrumentalist, Composer, Fav genre, Exploratory, Foreign languages, BPM, Frequency [Classical], Frequency [Country], Frequency [EDM], Frequency [Folk], Frequency [Gospel], Frequency [Hip hop], Frequency [Jazz], Frequency [K pop], Frequency [Latin], Frequency [Lofi], Frequency [Metal], Frequency [Pop], Frequency [R&B], Frequency [Rap], Frequency [Rock], Frequency [Video game music], Music effects, mental_scale]
Index: []

[0 rows x 28 columns]

we good to go no other values greater than 1000BPM

In [18]:
# BPM have many null values but we got limited data so just place mean value in those null values
# df["BPM"]=df["BPM"].fillna(value=df["BPM"].mean())

Spliting data

In [19]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

other preprocessing tasks such as removing unnecessary data, working with categorical data, feature engineering, and feature selection should be performed after splitting the data into training and test sets. This ensures that the test set remains independent and unbiased.

 if you want to fill missing values with the mean of a column, you should calculate the mean using only the training data and then use that value to fill missing values in both the training and test sets. This way, you avoid any information leakage from the test set into the training data1.

In [20]:
df

Age Primary streaming service  Hours per day While working  \
0    18.0                   Spotify            3.0           Yes   
1    63.0                   Pandora            1.5           Yes   
2    18.0                   Spotify            4.0            No   
3    61.0             YouTube Music            2.5           Yes   
4    18.0                   Spotify            4.0           Yes   
..    ...                       ...            ...           ...   
731  17.0                   Spotify            2.0           Yes   
732  18.0                   Spotify            1.0           Yes   
733  19.0   Other streaming service            6.0           Yes   
734  19.0                   Spotify            5.0           Yes   
735  29.0             YouTube Music            2.0           Yes   

    Instrumentalist Composer         Fav genre Exploratory Foreign languages  \
0               Yes      Yes             Latin         Yes               Yes   
1                No       No              Rock         Yes                No   
2                No       No  Video game music          No               Yes   
3                No      Yes              Jazz         Yes               Yes   
4                No       No               R&B         Yes                No   
..              ...      ...               ...         ...               ...   
731             Yes       No              Rock         Yes               Yes   
732             Yes       No               Pop         Yes               Yes   
733              No      Yes               Rap         Yes                No   
734             Yes       No         Classical          No                No   
735              No       No           Hip hop         Yes               Yes   

       BPM  ... Frequency [Latin] Frequency [Lofi] Frequency [Metal]  \
0    156.0  ...   Very frequently           Rarely             Never   
1    119.0  ...         Sometimes           Rarely             Never   
2    132.0  ...             Never        Sometimes         Sometimes   
3     84.0  ...   Very frequently        Sometimes             Never   
4    107.0  ...         Sometimes        Sometimes             Never   
..     ...  ...               ...              ...               ...   
731  120.0  ...         Sometimes           Rarely            Rarely   
732  160.0  ...             Never           Rarely             Never   
733  120.0  ...            Rarely        Sometimes            Rarely   
734  170.0  ...             Never            Never             Never   
735   98.0  ...             Never           Rarely             Never   

     Frequency [Pop]  Frequency [R&B]  Frequency [Rap] Frequency [Rock]  \
0    Very frequently        Sometimes  Very frequently            Never   
1          Sometimes        Sometimes           Rarely  Very frequently   
2             Rarely            Never           Rarely           Rarely   
3          Sometimes        Sometimes            Never            Never   
4          Sometimes  Very frequently  Very frequently            Never   
..               ...              ...              ...              ...   
731  Very frequently            Never           Rarely  Very frequently   
732  Very frequently            Never            Never        Sometimes   
733        Sometimes        Sometimes        Sometimes           Rarely   
734            Never            Never            Never            Never   
735        Sometimes  Very frequently  Very frequently  Very frequently   

    Frequency [Video game music] Music effects mental_scale  
0                      Sometimes           NaN         1.00  
1                         Rarely           NaN         3.00  
2                Very frequently     No effect         6.50  
3                          Never       Improve         5.50  
4                         Rarely       Improve         5.75  
..                           ...           ...          ...  
731                        N

In [21]:
df_features=df.drop("mental_scale",axis=1)
df_target=df["mental_scale"]

In [22]:
# df_features
# df_target

In [23]:
X_train,X_test,Y_train,Y_test=train_test_split(df_features,df_target,test_size=0.17,stratify=df["Fav genre"],random_state=19)

In [24]:
X_train.shape

(609, 27)

In [25]:
# separating cat and num columns
categorical_cols=X_train[X_train.columns[X_train.dtypes=="object"]].columns
numerical_cols=X_train[X_train.columns[X_train.dtypes!="object"]].columns

In [26]:
# X_train[categorical_cols[0]].mode()[0]

# handling null values

In [27]:
for cat in categorical_cols:
    imputer_value=X_train[cat].mode()[0] #values from train data only [frequent data]
    print(cat,imputer_value)
    X_train[cat]=X_train[cat].fillna(imputer_value)
    X_test[cat]=X_test[cat].fillna(imputer_value)

Primary streaming service Spotify
While working Yes
Instrumentalist No
Composer No
Fav genre Rock
Exploratory Yes
Foreign languages Yes
Frequency [Classical] Rarely
Frequency [Country] Never
Frequency [EDM] Never
Frequency [Folk] Never
Frequency [Gospel] Never
Frequency [Hip hop] Sometimes
Frequency [Jazz] Never
Frequency [K pop] Never
Frequency [Latin] Never
Frequency [Lofi] Never
Frequency [Metal] Never
Frequency [Pop] Very frequently
Frequency [R&B] Never
Frequency [Rap] Rarely
Frequency [Rock] Very frequently
Frequency [Video game music] Never
Music effects Improve


In [28]:
# similar for numerical data
for num in numerical_cols:
    imputer_num_value=X_train[num].mean() #values from train data only
    print(num,imputer_num_value)
    X_train[num]=X_train[num].fillna(imputer_num_value)
    X_test[num]= X_test[num].fillna(imputer_num_value)

Age 25.226600985221676
Hours per day 3.563300492610838
BPM 123.80152671755725


In [29]:
print(Y_test.isna().sum(),Y_train.isna().sum())
# no null in targets

0 0


In [30]:
# print(X_train.isna().sum(),X_test.isna().sum())
# handled all null values

In [31]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler

ohe = OneHotEncoder(dtype="int32",sparse_output=False)
ss=StandardScaler()

In [32]:
transformer=ColumnTransformer(transformers=[
    ('ohe',ohe,categorical_cols),
    ('ss',ss,numerical_cols)
],remainder="passthrough")

In [33]:
Xnew_train=transformer.fit_transform(X_train)

In [34]:
Xnew_train.shape,X_train.shape

((609, 102), (609, 27))

In [35]:
Xnew_test=transformer.transform(X_test)

In [36]:
Xnew_test.shape,X_test.shape

((125, 102), (125, 27))

In [37]:
# saving transformer
import pickle
with open('transformer.pkl','wb') as f:
    pickle.dump(transformer,f)


In [38]:
# from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet,LogisticRegression
# from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
# from sklearn.svm import SVR
# from sklearn.neighbors import KNeighborsRegressor
# from sklearn.neural_network import MLPRegressor
# from sklearn.tree import DecisionTreeRegressor
# from sklearn.gaussian_process import GaussianProcessRegressor
# from sklearn.ensemble import AdaBoostRegressor, BaggingRegressor

In [39]:
# sns.pairplot(df, hue="mental_scale")

# # Display the plot
# plt.show()

In [40]:
# objects
# linear_regression = LinearRegression()
# ridge_regression = Ridge()
# lasso_regression = Lasso()
# elastic_net = ElasticNet()
# random_forest = RandomForestRegressor()
# gradient_boosting = GradientBoostingRegressor()
# svr = SVR()
# k_neighbors = KNeighborsRegressor()
# neural_network = MLPRegressor()
# decision_tree = DecisionTreeRegressor()
# gaussian_process = GaussianProcessRegressor()
# adaboost = AdaBoostRegressor()
# bagging = BaggingRegressor()

In [41]:
# regressors_lists = [linear_regression, ridge_regression, lasso_regression, elastic_net,
#               random_forest, gradient_boosting, svr]

In [42]:
# test_results=[]
# for regressor in regressors_lists:
#     regressor.fit(Xnew_train, Y_train)
#     score = regressor.score(Xnew_test, Y_test) 
#     test_results.append((regressor,score))

In [43]:
# test_results

the fact as expected, this models has significantly low performance. Most trends I extracted in the EDA are the result of noisy data and do not scale well.

In [44]:
# we can try again change traget to categorical value
# Y_train.index
# Y_test.value_counts()

In [45]:
Y_train=Y_train.apply(lambda x: round(x))

In [46]:
Y_test=Y_test.apply(lambda x: round(x))

In [47]:
# ?classifiactions

from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier


In [48]:
gradient_boosting = GradientBoostingClassifier()


In [49]:
# acc_result
# [[LogisticRegression(max_iter=1000), 0.16],
 # [RandomForestClassifier(), 0.152],
 # [GradientBoostingClassifier(), 0.184],
 # [SVC(), 0.16],
 # [KNeighborsClassifier(), 0.128],
 # [GaussianNB(), 0.04],
 # [MLPClassifier(), 0.152],
 # [DecisionTreeClassifier(), 0.16],
 # [LinearDiscriminantAnalysis(), 0.144],
 # [AdaBoostClassifier(), 0.168],
 # [BaggingClassifier(), 0.16]]

In [50]:
gradient_boosting.fit(Xnew_train, Y_train)

GradientBoostingClassifier()

In [51]:
gradient_boosting.score(Xnew_test, Y_test)

0.176

In [52]:
predicted=gradient_boosting.predict(Xnew_test)

In [53]:
result_df=pd.DataFrame()
result_df["actual"]=Y_test
result_df["predicted"]=predicted
result_df

actual  predicted
650       8          4
417       5          5
275       5          4
100       4          5
242       3          6
..      ...        ...
653       6          6
60        3          0
647       2          3
512       1          5
314       4          6

[125 rows x 2 columns]

In [54]:

with open('model.pkl','wb') as f:
    pickle.dump(gradient_boosting,f)


In [55]:
model = pickle.load(open('model.pkl','rb'))

In [56]:
model.predict([Xnew_test[3]])

array([5], dtype=int64)

In [57]:
Y_test.iloc[3]

4

In [58]:
# testing with preprocessors
transformer_loaded = pickle.load(open('transformer.pkl','rb'))

In [59]:
data={
"Primary streaming service":["Spotify"],
"Age":[32.0],
"Hours per day":[12.0],
"While working":["Yes"],
"Instrumentalist":["Yes"],
"Composer":["Yes"],
"Fav genre":["Jazz"],
"Exploratory":["Yes"],
"Foreign languages":["Yes"],
"BPM":[200.0],
"Frequency [Classical]":["Never"],
"Frequency [Country]":["Never"],
"Frequency [EDM]":["Sometimes"],
"Frequency [Folk]":["Rarely"],
"Frequency [Gospel]":["Never"],
"Frequency [Hip hop]":["Sometimes"],
"Frequency [Jazz]":["Very frequently"],
"Frequency [K pop]":["Rarely"],
"Frequency [Latin]":["Never"],
"Frequency [Lofi]":["Sometimes"],
"Frequency [Metal]":["Never"],
"Frequency [Pop]":["Rarely"],
"Frequency [R&B]":["Sometimes"],
"Frequency [Rap]":["Rarely"],


"Frequency [Rock]":["Never"],
"Frequency [Video game music]":["Never"],
"Music effects":["Improve"]
}

In [63]:
testdf=pd.DataFrame(data)

In [65]:
testdf.shape

(1, 27)

In [69]:
X_test.head()

Age Primary streaming service  Hours per day While working  \
650  37.0                   Spotify            2.0           Yes   
417  32.0                   Spotify           12.0           Yes   
275  15.0                   Spotify            3.0            No   
100  18.0                   Spotify            2.0            No   
242  29.0                   Spotify            5.0           Yes   

    Instrumentalist Composer         Fav genre Exploratory Foreign languages  \
650              No       No              Rock         Yes                No   
417             Yes      Yes              Jazz         Yes               Yes   
275              No       No  Video game music          No                No   
100             Yes      Yes             Metal         Yes               Yes   
242             Yes      Yes         Classical         Yes               Yes   

            BPM  ... Frequency [K pop] Frequency [Latin] Frequency [Lofi]  \
650  100.000000  ...             Never             Never            Never   
417  200.000000  ...            Rarely             Never        Sometimes   
275   84.000000  ...             Never             Never  Very frequently   
100  149.000000  ...             Never             Never            Never   
242  123.801527  ...             Never             Never            Never   

    Frequency [Metal]  Frequency [Pop] Frequency [R&B] Frequency [Rap]  \
650         Sometimes        Sometimes          Rarely          Rarely   
417             Never           Rarely       Sometimes          Rarely   
275            Rarely  Very frequently           Never           Never   
100   Very frequently        Sometimes           Never          Rarely   
242            Rarely        Sometimes          Rarely           Never   

    Frequency [Rock] Frequency [Video game music] Music effects  
650        Sometimes                        Never        Worsen  
417            Never                        Never       Improve  
275  Very frequently              Very frequently     No effect  
100  Very frequently                       Rarely       Improve  
242        Sometimes                        Never       Improve  

[5 rows x 27 columns]